streaming/vision/convert/coco.py

```diff
@@ -0,0 +1,231 @@
+
+import json
+import os
+from argparse import ArgumentParser, Namespace
+from typing import Dict, Iterable
+
+import numpy as np
+import torch
+from PIL import Image
+from torch.utils.data import Dataset
+from tqdm import tqdm
+
+from streaming.base import MDSWriter
+from streaming.vision.convert.base import get_list_arg
+
+
+class _COCODetection(Dataset):
+    """PyTorch Dataset for the COCO dataset.
+
+    Args:
+        img_folder (str): Path to a COCO folder.
+        annotate_file (str): Path to a file that contains image id, annotations (e.g., bounding boxes and object
+            classes) etc.
+    """
+
+    def __init__(self, img_folder: str, annotate_file: str):
+        self.img_folder = img_folder
+        self.annotate_file = annotate_file
+
+        # Start processing annotation
+        with open(annotate_file) as fin:
+            self.data = json.load(fin)
+
+        self.images = {}
+
+        self.label_map = {}
+        self.label_info = {}
+        # 0 stands for the background
+        cnt = 0
+        self.label_info[cnt] = 'background'
+        for cat in self.data['categories']:
+            cnt += 1
+            self.label_map[cat['id']] = cnt
+            self.label_info[cnt] = cat['name']
+
+        # build inference for images
+        for img in self.data['images']:
+            img_id = img['id']
+            img_name = img['file_name']
+            img_size = (img['height'], img['width'])
+            if img_id in self.images:
+                raise Exception('dulpicated image record')
+            self.images[img_id] = (img_name, img_size, [])
+
+        # read bboxes
+        for bboxes in self.data['annotations']:
+            img_id = bboxes['image_id']
+            bbox = bboxes['bbox']
+            bbox_label = self.label_map[bboxes['category_id']]
+            self.images[img_id][2].append((bbox, bbox_label))
+
+        for k, v in list(self.images.items()):
+            if len(v[2]) == 0:
+                self.images.pop(k)
+
+        self.img_keys = list(self.images.keys())
+
+    #@property
+    def labelnum(self):
+        return len(self.label_info)
+
+    def __len__(self):
+        return len(self.images)
+
+    def __getitem__(self, idx: int):
+        img_id = self.img_keys[idx]
+        img_data = self.images[img_id]
+        fn = img_data[0]
+        img_path = os.path.join(self.img_folder, fn)
+
+        img = Image.open(img_path).convert('RGB')
+
+        htot, wtot = img_data[1]
+        bbox_sizes = []
+        bbox_labels = []
+
+        for (l, t, w, h), bbox_label in img_data[2]:
+            r = l + w
+            b = t + h
+            bbox_size = (l / wtot, t / htot, r / wtot, b / htot)
+            bbox_sizes.append(bbox_size)
+            bbox_labels.append(bbox_label)
+
+        bbox_sizes = torch.tensor(bbox_sizes)
+        bbox_labels = torch.tensor(bbox_labels)
+
+        return img, img_id, (htot, wtot), bbox_sizes, bbox_labels
+
+
+def parse_args() -> Namespace:
```

Comment:
i think this should go up top, above _COCODetection for ease of reading?

Reply:
Done in https://github.com/mosaicml/streaming/pull/5/commits/e1eaca7428674cb2941960d1ad7c39c7091c1b12.